pkg/storage/chunk/cache/lru_cache.go

```diff
@@ -0,0 +1,249 @@
+package cache
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"sync"
+	"unsafe"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/loki/pkg/logqlmodel/stats"
+	util_log "github.com/grafana/loki/pkg/util/log"
+	lru "github.com/hashicorp/golang-lru/simplelru"
+	"github.com/oklog/ulid"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+type codec string
+
+const (
+	codecHeaderSnappy             codec = "dvs" // As in "diff+varint+snappy".
+	codecHeaderSnappyWithMatchers codec = "dm"  // As in "dvs+matchers"
+)
+
+var DefaultLRUCacheConfig = LRUCacheConfig{
+	MaxSizeBytes: "250MB",
+}
+
+const maxInt = int(^uint(0) >> 1)
+
+const (
+	stringHeaderSize = 8
+	sliceHeaderSize  = 16
+)
+
+var ulidSize = uint64(len(ulid.ULID{}))
+
+type LRUCacheConfig struct {
+	MaxSizeBytes string `yaml:"max_size_bytes"`
+
+	Enabled bool `yaml:"enabled"`
+}
+
+// RegisterFlagsWithPrefix adds the flags required to config this to the given FlagSet
+func (cfg *LRUCacheConfig) RegisterFlagsWithPrefix(prefix, description string, f *flag.FlagSet) {
+	f.StringVar(&cfg.MaxSizeBytes, prefix+"fifocache.max-size-bytes", "500MB", description+"Maximum memory size of the cache in bytes. A unit suffix (KB, MB, GB) may be applied.")
+}
+
+func (cfg *LRUCacheConfig) Validate() error {
+	_, err := parsebytes(cfg.MaxSizeBytes)
+	return err
+}
+
+type LRUCache struct {
+	cacheType stats.CacheType
+
+	done chan struct{}
+
+	mtx sync.Mutex
+
+	logger           log.Logger
+	lru              *lru.LRU
+	maxSizeBytes     uint64
+	maxItemSizeBytes uint64
+
+	evicted     *prometheus.CounterVec
+	requests    *prometheus.CounterVec
+	hits        *prometheus.CounterVec
+	totalMisses prometheus.Counter
+	added       *prometheus.CounterVec
+	current     *prometheus.GaugeVec
+	bytesInUse  prometheus.Gauge
+	overflow    *prometheus.CounterVec
```

Comment:
I assume these metrics are used by all different cache implementations. Would it make sense to move them into a separate struct so they can be re-used?